### PR TITLE
Rabbitmq communication + graphql queries

### LINKED ISSUE #11: rabbitmq integration
- packet parsing
- reply/response helpers
- connection with rabbitmq
- broadcast + direct message support

### LINKED ISSUE #12: GraphQL loaders
loaders are a crucial part of a graphql api. it does batching and caching per request. this is to solve the N + 1 problem graphql normally has.

Loaders should be on the context - not as a resolver. an option is to use facebook's Dataloader npm package.

### PR DESCRIPTION
**Pull request by mrjvs**
in this PR:
 - rabbitMQ connecting
 - communication with the bot
 - graphql loaders
 - graphql queries for: mutual guilds, current discord user
 - simple caching for mutual guilds query
 - graphql context helpers

resolves #12
resolves #11

### CODE REVIEW

**Comment by mrjvs** [comment]
also @eddie-englund, this changes logic of the context. so your changes may be out of date after this merge

File: src/loaders/user.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  stdTTL: 5 * 60, // 5 minutes per user
+});
+
+interface DiscordGuild {
+  id: string;
+  iconUrl?: string;
+  onlineCount: number;
+  memberCount: number;
+  name: string;
+}
```
Comment (binaryoverload) [inline src/loaders/user.ts]: Why is this interface replicated from the types file?

File: src/loaders/user.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  mutualGuildCache.set(userId, flattedGuilds);
+  return flattedGuilds;
+}
+
+interface DiscordUser {
+  id: string;
+  name: string;
+  avatarUrl: string;
+  discriminator: string;
+}
```
Comment (binaryoverload) [inline src/loaders/user.ts]: Same here

File: src/modules/rabbitmq/helpers/meta.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { cascadeActions } from './actions';
+import { sendParsedDirectMessage } from '../parsed';
+
+const cacheExpiry = 60 * 1000; // 1 minute
+let cachedShardCount: number | null = null;
+let cacheTimeout: ReturnType<typeof setTimeout> | null = null;
+
+interface ShardCount {
+  shard_count: number;
+}
```
Comment (binaryoverload) [inline src/modules/rabbitmq/helpers/meta.ts:10]: Should these interfaces have a central location?

File: src/modules/rabbitmq/helpers/resources.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * send to and receive from resource queue
+ */
+export async function requestFromResources<T, K = Record<string, any>>(
+  action: cascadeActions,
+  options: Request<K>,
+): Promise<T> {
+  return await sendParsedDirectMessage<T>({
+    routingKey: `resource`,
```
Comment (binaryoverload) [inline src/modules/rabbitmq/helpers/resources.ts]:
```suggestion
    routingKey: 'resource',
```

File: src/modules/rabbitmq/helpers/resources.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * send to resource queue, no reply
+ */
+export async function sendToResources<K = Record<string, any>>(
+  action: cascadeActions,
+  options: Request<K>,
+): Promise<void> {
+  await sendParsedDirectMessage({
+    routingKey: `resource`,
```
Comment (binaryoverload) [inline src/modules/rabbitmq/helpers/resources.ts]:
```suggestion
    routingKey: 'resource',
```

File: src/modules/rabbitmq/index.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  const callbackQueueRes = await channel.assertQueue('', {
+    autoDelete: true,
+    exclusive: true,
+  });
+  callbackQueue = callbackQueueRes.queue;
+
+  startCallbackConsumer();
+
+  log.info(`Connected to rabbitMQ!`, { evt: 'success' });
```
Comment (binaryoverload) [inline src/modules/rabbitmq/index.ts]:
```suggestion
  log.info(`Connected to RabbitMQ!`, { evt: 'success' });
```

**Review by binaryoverload — CHANGES_REQUESTED** [review]
Looking good!

File: src/loaders/user.ts (hunk 1 above)
Comment (mrjvs) [reply src/loaders/user.ts]: ApiDiscordGuild is what's from the API (the bot)
DiscordGuild is what's used internally in the backend.
I want there to be explicit mapping between the two types, so the bot can change and we wouldn't have to change a bunch of types or change the graphql schema. we'd only change the Api version of the types

File: src/modules/rabbitmq/helpers/meta.ts (hunk 3 above)
Comment (mrjvs) [reply src/modules/rabbitmq/helpers/meta.ts:10]: probably not for shardcount as its very function specific. the other interfaces should be moved tho

File: src/config/parts/rabbitmq.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+export interface RabbitMQConf {
+  url: string;
+}
+
+export const rabbitMqConfSchema = joi.object({
+  url: joi.string().uri({
+    scheme: 'amqp',
+  }),
+});
```
Comment (weeryan17) [inline src/config/parts/rabbitmq.ts]: Add options for individual properties (like what we talked about in vc)

File: src/modules/rabbitmq/parsed.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  status_code: number;
+  error: undefined;
+  data: T;
+}
+
+interface IResponseError {
+  status_code: number;
+  error: ResponseErrorSub;
+  data: undefined;
+}
```
Comment (weeryan17) [inline src/modules/rabbitmq/parsed.ts]: Name this diffrently

**Review by weeryan17 — CHANGES_REQUESTED** [review]

**Review by binaryoverload — APPROVED** [review]

**Review by weeryan17 — APPROVED** [review]

**Pull request merged**